File: src/Cordango.Standalone/Notifications/Notifications.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) Cordango and contributors.
// Part of Cordango, the open application language and compiler: https://github.com/cordango/cordango
// Licensed under the Apache License, Version 2.0. See LICENSE in the repository root.

using System.Text.Json.Serialization;
using Cordango.Standalone.Data;
using Cordango.Standalone.Http;
using Cordango.Standalone.Records;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace Cordango.Standalone.Notifications;

/// <summary>
/// Something that happened, addressed to one person.
///
/// <para>In the application rather than in an inbox: a definition's <c>notify</c> effect means "tell
/// this person", and telling them inside the thing they are already using is the version that needs
/// no mail server, no deliverability and no unsubscribe link.</para>
/// </summary>
public sealed class Notification : IRecord
{
    public string Id { get; set; } = "";

    /// <summary>The directory Person this is for.</summary>
    [JsonPropertyName("person")] public string Person { get; set; } = "";

    [JsonPropertyName("title")] public string Title { get; set; } = "";
    [JsonPropertyName("message")] public string? Message { get; set; }

    /// <summary>Where to go when it is clicked — a record route, or nothing.</summary>
    [JsonPropertyName("link")] public string? Link { get; set; }

    [JsonPropertyName("created")] public DateTimeOffset Created { get; set; }
    [JsonPropertyName("read_at")] public DateTimeOffset? ReadAt { get; set; }
}

public static class NotificationModule
{
    public static ModelBuilder AddNotifications(this ModelBuilder builder)
    {
        ArgumentNullException.ThrowIfNull(builder);
        builder.Entity<Notification>(b =>
        {
            b.ToTable("notification");
            b.HasKey(e => e.Id);
            b.Property(e => e.Id).HasColumnName("id").HasMaxLength(64);
            b.Property(e => e.Person).HasColumnName("person").HasMaxLength(64).IsRequired();
            b.Property(e => e.Title).HasColumnName("title").IsRequired();
            b.Property(e => e.Message).HasColumnName("message");
            b.Property(e => e.Link).HasColumnName("link");
            b.Property(e => e.Created).HasColumnName("created");
            b.Property(e => e.ReadAt).HasColumnName("read_at");

            // Every read of this table is "what is waiting for me", newest first.
            b.HasIndex(e => new { e.Person, e.Created });
        });
        return builder;
    }
}

/// <summary>Writing one. Separate from the controller because effects call it from inside a command,
/// where there is no request to read.</summary>
public sealed class NotificationService
{
    private readonly CordangoDbContext _db;
    private readonly IClock _clock;
    private readonly IRecordIdGenerator _ids;

    public NotificationService(CordangoDbContext db, IClock clock, IRecordIdGenerator ids)
    {
        _db = db;
        _clock = clock;
        _ids = ids;
    }

    /// <summary>
    /// Tell somebody. Silently does nothing when there is nobody to tell.
    ///
    /// <para>That silence is deliberate and narrow: a <c>notify</c> whose recipient field is empty —
    /// an unassigned ticket, a claim with no approver — is a normal state of the data, not an error
    /// in the application. Failing the command because of it would make the command unrunnable
    /// exactly when it is most needed.</para>
    /// </summary>
    public async Task<bool> SendAsync(string? person, string title, string? message, string? link, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(person) || string.IsNullOrWhiteSpace(title)) return false;

        _db.Add(new Notification
        {
            Id = _ids.NewId(),
            Person = person,
            Title = title,
            Message = message,
            Link = link,
            Created = _clock.UtcNow,
        });

        await _db.SaveChangesAsync(ct);
        return true;
    }
}

/// <summary>What is waiting for the person asking, and nobody else.</summary>
[Route("api/notifications")]
public sealed class NotificationsController : ControllerBase
{
    private readonly CordangoDbContext _db;
    private readonly ICurrentUser _user;
    private readonly IClock _clock;

    public NotificationsController(CordangoDbContext db, ICurrentUser user, IClock clock)
    {
        _db = db;
        _user = user;
        _clock = clock;
    }

    [HttpGet]
    public async Task<IActionResult> List([FromQuery] bool unreadOnly = false, CancellationToken ct = default)
    {
        // Keyed by the signed-in person, so there is no id in the route to tamper with and no way to
        // phrase a request for somebody else's notifications.
        if (_user.PersonId is not { Length: > 0 } person)
            return Ok(new { items = Array.Empty<object>(), unread = 0 });

        var mine = _db.Set<Notification>().Where(n => n.Person == person);
        var unread = await mine.CountAsync(n => n.ReadAt == null, ct);

        var items = await (unreadOnly ? mine.Where(n => n.ReadAt == null) : mine)
            .OrderByDescending(n => n.Created)
            .Take(50)
            .ToListAsync(ct);

        return Ok(new { items, unread });
    }

    [HttpPost("{id}/read")]
    public async Task<IActionResult> Read(string id, CancellationToken ct)
    {
        if (_user.PersonId is not { Length: > 0 } person)
            return Unauthorized(this.Refuse("auth.required", "Sign in first."));

        // Scoped to the caller in the WHERE clause rather than checked afterwards: a notification
        // that is not theirs simply matches nothing.
        await _db.Set<Notification>()
            .Where(n => n.Id == id && n.Person == person)
            .ExecuteUpdateAsync(u => u.SetProperty(n => n.ReadAt, _clock.UtcNow), ct);

        return NoContent();
    }

    [HttpPost("read-all")]
    public async Task<IActionResult> ReadAll(CancellationToken ct)
    {
        if (_user.PersonId is not { Length: > 0 } person)
            return Unauthorized(this.Refuse("auth.required", "Sign in first."));

        await _db.Set<Notification>()
            .Where(n => n.Person == person && n.ReadAt == null)
            .ExecuteUpdateAsync(u => u.SetProperty(n => n.ReadAt, _clock.UtcNow), ct);

        return NoContent();
    }
}

File: src/Cordango.Standalone.Node/src/notifications/notifications.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) Cordango and contributors.
// Part of Cordango, the open application language and compiler: https://github.com/cordango/cordango
// Licensed under the Apache License, Version 2.0. See LICENSE in the repository root.

import { randomUUID } from "node:crypto";
import type { SqlDriver } from "../db/driver.js";

/**
 * Something that happened, addressed to one person.
 *
 * In the application rather than in an inbox: a definition's `notify` effect means "tell this
 * person", and telling them inside the thing they are already using is the version that needs no
 * mail server, no deliverability and no unsubscribe link.
 */
export interface Notification {
  id: string;
  /** The directory Person this is for. */
  person: string;
  title: string;
  message: string | null;
  /** Where to go when it is clicked — a record route, or nothing. */
  link: string | null;
  created: string;
  read_at: string | null;
}

const table = `
CREATE TABLE IF NOT EXISTS "notification" (
  "id" character varying(64) NOT NULL,
  "person" character varying(64) NOT NULL,
  "title" text NOT NULL,
  "message" text,
  "link" text,
  "created" timestamp with time zone NOT NULL,
  "read_at" timestamp with time zone,
  PRIMARY KEY ("id")
)`;

/** Every read of this table is "what is waiting for me", newest first. */
const index = `CREATE INDEX IF NOT EXISTS "ix_notification_person" ON "notification" ("person", "created")`;

type Row = Record<string, unknown>;

function text(value: unknown): string {
  return typeof value === "string" ? value : value === null || value === undefined ? "" : String(value);
}

function nullableText(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  if (value instanceof Date) return value.toISOString();
  return String(value);
}

function hydrate(row: Row): Notification {
  return {
    id: text(row["id"]),
    person: text(row["person"]),
    title: text(row["title"]),
    message: nullableText(row["message"]),
    link: nullableText(row["link"]),
    created: nullableText(row["created"]) ?? new Date(0).toISOString(),
    read_at: nullableText(row["read_at"]),
  };
}

/**
 * Writing one, and reading your own.
 *
 * Separate from the route because effects call it from inside a command, where there is no request
 * to read.
 */
export class NotificationService {
  constructor(private readonly driver: SqlDriver) {}

  async ensureTables(): Promise<void> {
    await this.driver.query(table);
    await this.driver.query(index);
  }

  /**
   * Tell somebody. Silently does nothing when there is nobody to tell.
   *
   * A `notify` whose recipient token resolved to nothing — an unassigned ticket, a claim with no
   * approver yet — is a notification with no addressee, and writing it would put a row in the table
   * that nobody will ever see and nobody can clear. The definition said "tell the assignee"; there
   * is no assignee; the honest outcome is silence.
   */
  async send(
    person: string | null | undefined,
    title: string,
    message: string | null,
    link: string | null,
  ): Promise<void> {
    if (person === null || person === undefined || person.trim() === "") return;
    if (title.trim() === "") return;

    await this.driver.execute(
      `INSERT INTO "notification" ("id","person","title","message","link","created","read_at")
       VALUES ($1,$2,$3,$4,$5,$6,NULL)`,
      [randomUUID().replaceAll("-", ""), person.trim(), title, message, link, new Date()],
    );
  }

  /** This person's notifications, newest first. Never anybody else's: the person is taken from the
   * session, never from the query string. */
  async list(person: string, take = 50): Promise<Notification[]> {
    const rows = await this.driver.query(
      `SELECT * FROM "notification" WHERE "person" = $1 ORDER BY "created" DESC LIMIT $2`,
      [person, Math.min(Math.max(take, 1), 200)],
    );
    return rows.map((row) => hydrate(row as Row));
  }

  async unreadCount(person: string): Promise<number> {
    const rows = await this.driver.query(
      `SELECT COUNT(*) AS "count" FROM "notification" WHERE "person" = $1 AND "read_at" IS NULL`,
      [person],
    );
    const value = (rows[0] as Row | undefined)?.["count"];
    return typeof value === "number" ? value : Number(text(value) || "0");
  }

  /** Mark one read, if it is this person's. The person is part of the WHERE rather than checked
   * afterwards, so marking somebody else's notification read is not expressible. */
  async markRead(person: string, id: string): Promise<boolean> {
    const affected = await this.driver.execute(
      `UPDATE "notification" SET "read_at" = $3 WHERE "id" = $1 AND "person" = $2 AND "read_at" IS NULL`,
      [id, person, new Date()],
    );
    return affected > 0;
  }

  async markAllRead(person: string): Promise<number> {
    return this.driver.execute(
      `UPDATE "notification" SET "read_at" = $2 WHERE "person" = $1 AND "read_at" IS NULL`,
      [person, new Date()],
    );
  }
}
